src/gc.c

```diff
@@ -1461,86 +1474,113 @@ static int push_root(jl_ptls_t ptls, jl_value_t *v, int d, int bits)
             }
         }
     }
-    else if (vt == (jl_value_t*)jl_module_type) {
+    else if (vt == jl_module_type) {
         // should increase nptr here
-        bits = gc_setmark(ptls, v, sizeof(jl_module_t));
-        if (d >= MAX_MARK_DEPTH)
-            goto queue_the_root;
-        refyoung |= gc_mark_module(ptls, (jl_module_t*)v, d);
+        refyoung |= gc_mark_module(ptls, (jl_module_t*)v, d, bits);
     }
-    else if (vt == (jl_value_t*)jl_task_type) {
+    else if (vt == jl_task_type) {
         // ditto nptr
-        bits = gc_setmark(ptls, v, sizeof(jl_task_t));
-        if (d >= MAX_MARK_DEPTH)
-            goto queue_the_root;
-        gc_mark_task(ptls, (jl_task_t*)v, d);
+        gc_mark_task(ptls, (jl_task_t*)v, d, bits);
         // tasks should always be remarked since we do not trigger the write barrier
         // for stores to stack slots
         refyoung = 1;
     }
-    else if (vt == (jl_value_t*)jl_symbol_type) {
-        // symbols have their own allocator and are never freed
-    }
-    // this check should not be needed but it helps catching corruptions early
-    else if (jl_typeof(vt) == (jl_value_t*)jl_datatype_type) {
-        jl_datatype_t *dt = (jl_datatype_t*)vt;
-        size_t dtsz = jl_datatype_size(dt);
-        bits = gc_setmark(ptls, v, dtsz);
-        if (d >= MAX_MARK_DEPTH)
-            goto queue_the_root;
-
-        int nf = (int)jl_datatype_nfields(dt);
-        // TODO check if there is a perf improvement for objects with a lot of fields
-        // int fdsz = sizeof(void*)*nf;
-        // void** children = alloca(fdsz);
-        // int ci = 0;
+    else {
+        int nf = (int)jl_datatype_nfields(vt);
         for(int i=0; i < nf; i++) {
-            if (jl_field_isptr(dt, i)) {
+            if (jl_field_isptr(vt, i)) {
                 nptr++;
                 jl_value_t **slot = (jl_value_t**)((char*)v +
-                                                   jl_field_offset(dt, i));
+                                                   jl_field_offset(vt, i));
                 jl_value_t *fld = *slot;
                 if (fld) {
                     verify_parent2("object", v, slot, "field(%d)", i);
-                    //children[ci++] = fld;
                     refyoung |= gc_push_root(ptls, fld, d);
                 }
             }
         }
-        //while(ci)
-        //  refyoung |= gc_push_root(ptls, children[--ci], d);
-    }
-    else {
-        jl_printf(JL_STDOUT, "GC error (probable corruption) :\n");
```

Comment:
seems like it would be useful to preserve this branch in `gc_mark_obj`?
```
// this check should not be needed but it helps catching corruptions early
```